### PR TITLE
Update cover image markup and CSS

### PR DESCRIPTION
**Pull request by samikeijonen**
## Description
Update cover image markup and CSS discussed in #2902.

## How Has This Been Tested?
- Tested with default themes and random .org themes.
- **Running `npm test` gives error**. I can't seem to track down why? It's about `"isValid": false,`.

```php
 File 'core__cover-image.json' does not match expected value:

    expect(received).toEqual(expected)

    Expected value to equal:
      [{"attributes": {"dimRatio": 50, "hasParallax": false, "title": ["Guten Berg!"], "url": "https://cldup.com/uuUqE_dXzy.jpg"}, "isValid": true, "name": "core/cover-image", "originalContent": "<div class=\"wp-block-cover-image has-background-dim-50 has-background-dim\" style=\"background-image:url(https://cldup.com/uuUqE_dXzy.jpg)\">
        <p class=\"wp-block-cover-image-text\">Guten Berg!</p>
    </div>", "uid": "_uid_0"}]
    Received:
      [{"attributes": {"dimRatio": 50, "hasParallax": false, "title": ["Guten Berg!"], "url": "https://cldup.com/uuUqE_dXzy.jpg"}, "isValid": false, "name": "core/cover-image", "originalContent": "<div class=\"wp-block-cover-image has-background-dim-50 has-background-dim\" style=\"background-image:url(https://cldup.com/uuUqE_dXzy.jpg)\">
        <p class=\"wp-block-cover-image-text\">Guten Berg!</p>
    </div>", "uid": "_uid_0"}]

    Difference:

    - Expected
    + Received

    @@ -6,11 +6,11 @@
            "title": Array [
              "Guten Berg!",
            ],
            "url": "https://cldup.com/uuUqE_dXzy.jpg",
          },
    -     "isValid": true,
    +     "isValid": false,
```

## Types of changes
- Change `<section>` to `<div>`.
- Change hardcoded `<h2>` to `<p>`.
- Add class `<p class="wp-block-cover-image-text">`. This was mainly for easier styling in the editor.
- Note that I don't how to correctly add class `wp-block-cover-image-text`. I have now added it like this: `<p className={ classnames( 'wp-block-cover-image-text' ) }>`

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [ ] My code follows has proper inline documentation.

### CODE REVIEW

**Comment by aduth** [comment]
Related: #2541

**Comment by samikeijonen** [comment]
In 96e5b504881961fc861c2142dfc0a52679c1c8c3 I added align classes mentioned in issue #3516.

**Comment by afercia** [comment]
@aduth  when you have a chance: what is the best way to help this PR move on? The changes proposed here make a lot of sense to me.

**Comment by aduth** [comment]
With #3665 having been merged, we now have an upgrade path for block markup.

#3851 is another example of a migration.

If we want to avoid breakage for existing blocks, I suggest refactoring to take advantage of this new mechanism.

**Comment by samikeijonen** [comment]
1. I added deprecate version but not sure if it's working.
1. When running `npm test` I still get errors. Something about `File 'core__pullquote__multi-paragraph.json' does not match expected value:`

**Comment by samikeijonen** [comment]
Should `wp-block-cover-image-text` class be `wp-block-cover-image__text`?

File: blocks/library/cover-image/index.js [hunk 1]
```diff
@@ -217,11 +221,44 @@ registerBlockType( 'core/cover-image', {
...
+				selector: 'h2',
+			},
+		},
+
+		save( { attributes, className } ) {
+			const { url, title, hasParallax, dimRatio, align } = attributes;
+			const style = url ?
+				{ backgroundImage: `url(${ url })` } :
+				undefined;
+			const classes = classnames(
```
Comment (aduth) [inline blocks/library/cover-image/index.js]: It's possible we could [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) this up a bit by moving the common logic into a function (similar to the shared `blockAttributes` variable).

**Review by aduth — COMMENTED** [review]
>Should wp-block-cover-image-text class be wp-block-cover-image__text?

The general guidelines are outlined here:

https://github.com/WordPress/gutenberg/blob/master/docs/coding-guidelines.md#naming

In the case of blocks (particularly block front-end styling), the rules haven't been quite consistent with this. Based on the generated class I think `.wp-block-cover-image-text` could be fine enough.

The alternative being that we don't assign a class to the paragraph and instead apply styles via an element selector (`.wp-block-cover-image > p`), but arguments can be made discouraging this, as it makes the markup less tolerant to deprecation changes.

Speaking of styling and deprecation, one important consideration we should have with all blocks, including the changes here, is that we should anticipate that some older blocks exist with the deprecated markup (here, the `h2` tag) and therefore we ought to always keep those styles around so users' legacy blocks' don't suddenly start breaking when viewed on the front-end (cc @youknowriad ).

>When running npm test I still get errors. Something about File 'core__pullquote__multi-paragraph.json' does not match expected value:

Have you tried running the npm script for regenerating fixtures?

```
npm run fixtures:regenerate
```

**Comment by aduth** [comment]
>Speaking of styling and deprecation, one important consideration we should have with all blocks, including the changes here, is that we should anticipate that some older blocks exist with the deprecated markup (here, the h2 tag) and therefore we ought to always keep those styles around so users' legacy blocks' don't suddenly start breaking when viewed on the front-end (cc @youknowriad ).

Related: #4408, https://github.com/WordPress/gutenberg/pull/3851#discussion_r157050519

**Comment by samikeijonen** [comment]
Hmmm I messed my branch when accidently deleting also this branch on local. Now I only have one commit message in here: fc6b77adbbef305cf040b8ea7482cd32d84fedc5

In commit 3690713c467c6c41dce297f706ee1ea22bde408a I have added deprecate styles for `<h2>`.

File: blocks/library/cover-image/index.js (hunk 1 above)
Comment (samikeijonen) [reply blocks/library/cover-image/index.js]: Only for `const style`? I can do the similar as `function dimRatioToClass()`.

**Comment by jasmussen** [comment]
Does this PR also ensure that no empty `<p>` tags are output if the user hasn't entered any title? It's not a blocker, but if it's easy then we should do it.

**Comment by samikeijonen** [comment]
@jasmussen I just tested and it does output empty `<p>`. I can try to look into it.

If @aduth have time to push me into right direction that would be nice.

**Comment by youknowriad** [comment]
Not sure what's left on this PR. How can we help?

**Comment by samikeijonen** [comment]
@youknowriad Pretty much nothing left to do if you ask me:) There seems to be some merge conflicts, I'll try to look into them over the weekend.

**Comment by samikeijonen** [comment]
Fixed merge conflict and added link color to white.

File: blocks/library/cover-image/index.js [hunk 2]
```diff
@@ -210,17 +211,16 @@ export const settings = {
...
 
 	save( { attributes, className } ) {
 		const { url, title, hasParallax, dimRatio, align, contentAlign } = attributes;
-		const style = url ?
-			{ backgroundImage: `url(${ url })` } :
-			undefined;
+		const style = backgroundImageStyles( url );
 		const classes = classnames(
 			className,
+			align ? `align${ align }` : null,
```
Comment (youknowriad) [inline blocks/library/cover-image/index.js]: This seems like a duplicate?

**Review by youknowriad — COMMENTED** [review]
This looks good to me, there's a small duplicate and failing test. I'll merge once fixed

File: blocks/library/cover-image/index.js (hunk 2 above)
Comment (samikeijonen) [reply blocks/library/cover-image/index.js]: I removed duplicate and fixed merge conflict.

**Comment by samikeijonen** [comment]
@youknowriad Any update on this one? There is check errors but other that should be good to go.

I'm going trough all the changes in the front-end and this popped up.

**Comment by youknowriad** [comment]
Actually, I was waiting for you to fix the tests before merging :). Do you want some help in doing so?

**Comment by samikeijonen** [comment]
Yes please:) I don't know how to fix the tests, tried couple of times but don't really understand the reasons.

**Comment by youknowriad** [comment]
@jasmussen Can I have a quick design review for this before merge. The text looks lighter because it's not an H2 anymore but maybe it's better this way?

**Comment by samikeijonen** [comment]
Thanks for the fix!

We could add `font-weight: 700`.

**Comment by samikeijonen** [comment]
@jasmussen Can we merge this?

**Comment by jasmussen** [comment]
Ack my apologies, I don't know how I missed the previous ping! 

I got these two screenshots from the branch:

screen shot 2018-03-08 at 14 13 36

screen shot 2018-03-08 at 14 13 44

If they look right to you, then yes, this is a good look I think. 👍 👍

**Comment by samikeijonen** [comment]
Looks right to me, but I didn't add `font-weight: 700`. I'd say let's merge this now so we have correct markup. Then we can create minor design fixes if needed.

**Comment by samikeijonen** [comment]
@youknowriad Can we merge this?

**Pull request merged**